flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/recovery/RescalingStreamTaskNetworkInput.java

```diff
@@ -232,7 +232,9 @@ public Predicate<StreamRecord<T>> apply(InputChannelInfo channelInfo) {
                             channelInfo.getGateIdx(), this::createPartitioner);
             // use a copy of partitioner to ensure that the filter of ambiguous virtual channels
             // have the same state across several subtasks
-            return new RecordFilter<>(partitioner.copy(), inputSerializer, subtaskIndex);
+            StreamPartitioner<T> partitionerCopy = partitioner.copy();
+            partitionerCopy.setup(numberOfChannels);
+            return new RecordFilter<>(partitionerCopy, inputSerializer, subtaskIndex);
```

Comment:
Why do we need this change? 🤔

Reply:
By testing reported bug, I found little problem with copying partitioner in rescaling case. It can be throws ArithmeticException /by zero for CustomPartitioners (there used something like ` ... % numberOfChannels`), because after copying we have numberOfChannels is 0.

Reply:
Ok, in that case: is this a production bug (sounds like it), or only issue in testing?

1. If this is a production bug, please:
  - create a separate jira ticket for this issue
  - extract this change to a separate commit (can be part of this PR or a new PR, as you prefer) 
  - add a test coverage?
2. If that's just testing issue:
  - extract this change to a separate `[hotfix]` commit in this PR

Reply:
Extracted to `[hotfix]` [PR](https://github.com/apache/flink/pull/24811)